src/fu-security-attr.c

```diff
@@ -301,17 +305,281 @@ fu_security_attrs_to_json(FuSecurityAttrs *attrs, JsonBuilder *builder)
+ *		"removed": {
+ *		...
+ *		}
+ *	}
+ * }
+ *
+ * Returns: TRUE on success and FALSE on error.
+ *
+ * Since: 1.7.0
+ *
+ */
+static gboolean
+fu_security_attr_deep_object_compare(FwupdSecurityAttr *current_attr,
+				     JsonObject *previous_json_obj,
+				     JsonBuilder *result_builder)
+{
+	g_autoptr(GPtrArray) flag_array = NULL;
+	if (previous_json_obj != NULL) {
+		if (fwupd_security_attr_get_level(current_attr) ==
+		    json_object_get_int_member(previous_json_obj, FWUPD_RESULT_KEY_HSI_LEVEL))
+			return TRUE;
+
+		json_builder_set_member_name(
+		    result_builder,
+		    json_object_get_string_member(previous_json_obj,
+						  FWUPD_RESULT_KEY_APPSTREAM_ID));
+		json_builder_begin_object(result_builder);
+		json_builder_set_member_name(result_builder, "previous");
+		json_builder_begin_object(result_builder);
+		fu_security_attr_dup_json(previous_json_obj, result_builder);
+		json_builder_end_object(result_builder);
+		json_builder_set_member_name(result_builder, "current");
+	} else {
+		json_builder_set_member_name(result_builder,
+					     fwupd_security_attr_get_appstream_id(current_attr));
+		json_builder_begin_object(result_builder);
+		json_builder_set_member_name(result_builder, "new");
+	}
+	json_builder_begin_object(result_builder);
+	json_builder_set_member_name(result_builder, FWUPD_RESULT_KEY_HSI_LEVEL);
+	json_builder_add_int_value(result_builder, fwupd_security_attr_get_level(current_attr));
+	json_builder_set_member_name(result_builder, FWUPD_RESULT_KEY_HSI_RESULT);
+	json_builder_add_string_value(
+	    result_builder,
+	    fwupd_security_attr_result_to_string(fwupd_security_attr_get_result(current_attr)));
+	json_builder_set_member_name(result_builder, FWUPD_RESULT_KEY_NAME);
+	json_builder_add_string_value(result_builder, fwupd_security_attr_get_name(current_attr));
+	flag_array = fu_security_attr_flag_to_string_array(current_attr);
+	if (flag_array != NULL) {
+		json_builder_set_member_name(result_builder, FWUPD_RESULT_KEY_FLAGS);
+		json_builder_begin_array(result_builder);
+		for (guint i = 0; i < flag_array->len; i++) {
+			json_builder_add_string_value(
+			    result_builder,
+			    // g_strdup(g_ptr_array_index(flag_array, i)));
+			    g_ptr_array_index(flag_array, i));
+		}
+		json_builder_end_array(result_builder);
+	}
+	json_builder_end_object(result_builder);
+	json_builder_end_object(result_builder);
+
+	return FALSE;
+}
+
+static void
+fu_security_attr_append_remove_to_result(JsonObject *previous_json_obj, JsonBuilder *result_builder)
+{
+	json_builder_set_member_name(
+	    result_builder,
+	    json_object_get_string_member(previous_json_obj, FWUPD_RESULT_KEY_APPSTREAM_ID));
+	json_builder_begin_object(result_builder);
+	json_builder_set_member_name(result_builder, "removed");
+	json_builder_begin_object(result_builder);
+	fu_security_attr_dup_json(previous_json_obj, result_builder);
+	json_builder_end_object(result_builder);
+	json_builder_end_object(result_builder);
+}
+
+gchar *
```

Comment:
i think some function documentation would help a lot here, i.e. what's it *doing*?

Reply:
I'll put some descriptions in the function document. Thanks for pointing this out.